implement bookmark copying & ZCP bookmark support

problame opened the pull request:
This feature allows copying existing bookmarks using
```
  zfs bookmark fs#target fs#newbookmark
```
There are some niche use cases for such functionality,
e.g. when using bookmarks as markers for replication progress.

Copying redaction bookmarks produces a normal bookmark that
cannot be used for redacted send (we are not duplicating
the redaction object).

ZCP support for bookmarking (both creation and copying) is implemented in a separate commit in this PR.

Overview:

- Terminology:
    - source = existing snapshot or bookmark
    - new/bmark = new bookmark
- Implement bookmark copying in `dsl_bookmark.c`
  - create new bookmark node
  - copy source's `zbn_phys` to new's `zbn_phys`
  - zero-out redaction object id in copy
- Extend existing bookmark ioctl nvlist schema to accept
  bookmarks as sources
  - => `dsl_bookmark_create_nvl_validate` is authoritative
- use `dsl_dataset_is_before` check for both snapshot
  and bookmark sources
- Adjust CLI
  - refactor shortname expansion logic in `zfs_do_bookmark`
- Update man pages
  - warn about redaction bookmark handling
- Add test cases
  - CLI
  - ZCP (very basic)
  - pyyzfs libzfs_core bindings

### How Has This Been Tested?

- The tests added as part of this PR
- Manual testing on a file-backed pool, unencrypted dataset

### Types of changes
- ~~Bug fix (non-breaking change which fixes an issue)~~
- [x] New feature (non-breaking change which adds functionality)
-  ~~Performance enhancement (non-breaking change which improves efficiency)~~
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- ~~Breaking change (fix or feature that would cause existing functionality to change)~~
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

Ornias1993 commented on cmd/zfs/zfs_main.c:
```diff
@@ -7471,16 +7471,17 @@ zfs_do_diff(int argc, char **argv)
  *
- * Creates a bookmark with the given name from the given snapshot.
+ * Creates a bookmark with the given name from the target snapshot
+ * or creates a copy of an existing target bookmark.
```
I personally don't like the use of the term "Target", "given snapshot" is clear to everyone.
"Target" me be interprented the other way around (as in: Goal), due to translation errors.

Ornias1993 commented on man/man8/zfs.8:
```diff
@@ -3607,9 +3607,9 @@ The command can also be given a path to a ZFS file system shared on the system.
 .Nm
 .Cm bookmark
-.Ar snapshot bookmark
+.Ar snapshot Ns | Ns Ar bookmark bookmark
```
I take you mean "Origin bookmark" and "target bookmark" here for the copy function?
Maybe clear that up, instead of 2 times "bookmark"?

problame commented on module/zfs/dsl_bookmark.c:
```diff
@@ -413,6 +427,148 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
+
+	// TODO need to take care of take care of redaction bookmarks?
+
+	if (new_dbn->dbn_phys.zbm_flags & ZBM_FLAG_HAS_FBN) {
```
`dsl_bookmark_node_add` already increments `bookmark_v2`

problame commented on cmd/zfs/zfs_main.c:
```diff
@@ -7500,59 +7501,82 @@ zfs_do_bookmark(int argc, char **argv)
 	/* check number of arguments */
 	if (argc < 1) {
-		(void) fprintf(stderr, gettext("missing snapshot argument\n"));
+		(void) fprintf(stderr, gettext("missing target argument\n"));
```
I chose the `new` and `target` naming because the nvlist that is passed to the kernel associates `new => existing`.
But sure, `source` is less confusing because most users will think about it the other way around (I did too before learning about the nvlist structure).

How do we deal with the man page then, since the `source` (new term) now cannot only be a `snapshot` but also a `bookmark`? Also see https://github.com/zfsonlinux/zfs/pull/9571#discussion_r344594336 @Ornias1993

problame replied:
Will go with `source`

problame commented on module/zfs/dsl_bookmark.c:
```diff
@@ -413,6 +427,148 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
+	dsl_bookmark_node_t *new_dbn = dsl_bookmark_node_alloc(new_shortname);
+	dsl_bookmark_copy_phys(&new_dbn->dbn_phys, &target_phys);
+
+	// TODO need to take care of take care of redaction bookmarks?
```
Do we really want to copy the redaction object?

AFAICT it's immutable once it is created through `zfs redact`, and the redaction bookmark doesn't exist until after `zfs redact` completes,  which means we cannot come into a situation where we would try to copy an incomplete redaction bookmark.

=> can't we just have a reference count for this? (obviously, for O(1) bookmark deletion, this would need to be tracked on disk

pcd1193182 replied:
That should be fine as well.

problame commented on module/zfs/dsl_bookmark.c:
```diff
@@ -413,6 +427,148 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
+	int error;
+	zfs_bookmark_phys_t bmark_phys = { 0 };
+
+	/* Verify that the new bookmark does not already exist */
```
Good catch, no, the clone code didn't.
I'll add some tests for this, too.

BTW: the `create` code does the check, but only very implicitly because `dsl_bookmark_create_check_impl` does the bookmark-existence check based on the shortname, relative to the snapshot's dataset:

https://github.com/zfsonlinux/zfs/blob/03b516cf864230375dd5077c4792716322895010/module/zfs/dsl_bookmark.c#L140-L160

Is such implicit checking considered good enough for the DSL, or should I add the checks I'll add to the bookmark clone path to the bookmark create path as well?

problame replied:
( Code block for this is `/* Verify that new and target bookmark are on the same dataset */` above )

pcd1193182 replied:
I think that implicit checking is fine, for the create path. Might be good to add a comment that calls out that it does that, in case some future refactor accidentally breaks it.

ahrens replied:
The relevant check in dsl_bookmark_create_check_impl() is dsl_dataset_is_before(), which I would say is relatively explicit.  It's comment says, `Return TRUE if 'earlier' is an earlier snapshot in 'later's timeline.`

problame commented on module/zfs/dsl_bookmark.c:
```diff
@@ -413,6 +427,148 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
+
+	for (nvpair_t *pair = nvlist_next_nvpair(dbcc->dbcc_bmarks, NULL);
+	    pair != NULL; pair = nvlist_next_nvpair(dbcc->dbcc_bmarks, pair)) {
+		dsl_bookmark_clone_sync_impl(nvpair_name(pair),
```
It was a copy-paste leftover from from the `_create` code

problame commented on module/zfs/dsl_bookmark.c:
```diff
@@ -413,6 +420,194 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
+	VERIFY0(dsl_bookmark_hold_ds(dp, new_name, &bmark_fs_new, FTAG,
+	    &new_shortname));
+	/* Bookmarks must be on the same dataset */
+	VERIFY3U(bmark_fs_target->ds_objset, ==, bmark_fs_new->ds_objset);
```
Is this a legitimate and correct way to assert identity of two datasets? Didn't find a `dsl_dataset_equals` function or similar.

problame replied:
hat should be `ds_object`, but the question hasn't changed

pcd1193182 replied:
Verifying that they have the same ds_object will work.

problame commented on module/zfs/dsl_bookmark.c:
```diff
@@ -413,6 +420,194 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
+	if (target_redacted) {
+		// TODO copy zbm_redaction_obj and adjust new_dbn->dbn_phys.zbm_redaction_obj
+		// or use reference counting (the redaction object should be immutable if the
+		// bookmark exists, right?)
```
@pcd1193182  what do you think about this?

problame replied:
* bookmark is present and appears in zfs list even before the redaction is complete
* (i.e. concurrent to zfs redact, it's possible to zfs list and observe the redaction bookmark despite ongoing redaction)

When to know when redaction is complete? => Slack:

> it's when rbp_object = rbp_blkid = UINT64_MAX
> that's when the list is known to be complete

pcd1193182 replied:
I would have to run some tests, I don't recall for sure if the bookmark is visible before the redaction list is completed. If it is, the semantics of copying it are not immediately obvious to me, though if it were defined to just reference the existing redaction list that will get updated when the `zfs redact` command finishes, I think that would be ok.

problame replied:
I updated the TODO comment in the code to reflect all alternatives.

problame commented:
I addressed most of the review comments. Open issues:

- [ ] mechanism to copy redaction bookmarks
- [ ] naming: `target` vs `src` vs `origin`
- [ ] channel program support (is this necessary?)

problame commented on tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.bookmark.create.ksh:
```diff
@@ -0,0 +1,43 @@
+#
+
+#
+# Copyright (c) 2016, 2017 by Delphix. All rights reserved.
```
fix this

problame commented on tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.bookmark.clone.ksh:
```diff
@@ -0,0 +1,45 @@
+#
+
+#
+# Copyright (c) 2016, 2017 by Delphix. All rights reserved.
```
fix this

problame commented on tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.bookmark.create.ksh:
```diff
@@ -0,0 +1,43 @@
+. $STF_SUITE/tests/functional/channel_program/channel_common.kshlib
+
+#
+# DESCRIPTION: Make sure basic snapshot functionality works in channel programs
```
fix this

problame commented on tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.bookmark.clone.ksh:
```diff
@@ -0,0 +1,45 @@
+. $STF_SUITE/tests/functional/channel_program/channel_common.kshlib
+
+#
+# DESCRIPTION: Make sure basic snapshot functionality works in channel programs
```
fix this

problame commented:
We talked about this PR on the OpenZFS montly call. From the notes:

* redaction object might be too large to copy it in syncing context, would have to be done in the background
* => don’t implement redaction bookmark cloning at this point
* => fail on: zfs bookmark #redactionbook #newbook
* => create thin bookmark:  zfs bookmark --thin #redactionbook #newbook
will not contain the redaction object itself

ahrens commented on cmd/zfs/zfs_main.c:
```diff
@@ -7500,59 +7518,86 @@ zfs_do_bookmark(int argc, char **argv)
+		if (source_short == source &&
+		    bookname_short == bookname) {
+			gettext("either source or bookmark must be specified "
+			    "as full dataset paths");
```
I think we want to print this out.

ahrens commented on module/zfs/zfs_ioctl.c:
```diff
@@ -3651,6 +3681,48 @@ zfs_ioc_bookmark(const char *poolname, nvlist_t *innvl, nvlist_t *outnvl)
+
+/* ARGSUSED */
+static int
+zfs_ioc_bookmark_clone(const char *poolname, nvlist_t *innvl, nvlist_t *outnvl)
```
What's the reasoning behind creating a new ioctl, vs. allowing additional types of arguments (source can be a bookmark) to the existing ioctl?

Regardless of the user-kernel interface, I think that we should be able to reuse more of the zfs_ioc_bookmark() code for the new functionality.  i.e. modify `dsl_bookmark_create_*` to handle the source being a bookmark.  You would probably make it convert the source snapshot to a bookmark_phys_t, as we do for zfs send.  See handling of `ancestor_zb` in `dmu_send()`:

```
		zfs_bookmark_phys_t *zb = &dspp.ancestor_zb;
		if (strchr(fromsnap, '@') != NULL) {
			err = dsl_dataset_hold(dspp.dp, fromsnap, FTAG,
			    &fromds);
				if (!dsl_dataset_is_before(dspp.to_ds, fromds,
				    0)) {
					err = SET_ERROR(EXDEV);
				} else {
					ASSERT3U(dspp.is_clone, ==,
					    (dspp.to_ds->ds_dir !=
					    fromds->ds_dir));
					zb->zbm_creation_txg =
					    dsl_dataset_phys(fromds)->
					    ds_creation_txg;
...
		} else {
			dspp.numfromredactsnaps = NUM_SNAPS_NOT_REDACTED;
			err = dsl_bookmark_lookup(dspp.dp, fromsnap, dspp.to_ds,
			    zb);

```

problame replied:
Both the decision to create a new ioctl and the decision to have completely separate code paths for creation vs cloning were made because

* I wasn't clear about how much overlap there actually is between source_type=snapshot vs source_type=bookmark
* I did not know whether people would like the feature to live in the `zfs clone` vs `zfs bookmark` subcommand
* type-dependent branching deep down in the check tree makes it harder to reason about the correctness of the checks due to the increased amount of possible control flows

However, I see the point that a lot of the `dsl_bookmark_clone_*` stuff is boilerplate.
Thanks for the hint to the `zfs send` code.
I will merge the cloning functionality into the `dsl_bookmark_create_*` machinery.

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -413,6 +456,219 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
+			return (SET_ERROR(EINVAL));
+	}
+
+	/* Verify that new and source bookmark are on the same dataset */
```
They shouldn't have to be on the same dataset.  Like normal bookmark creation, the source can be any point in the history of the filesystem that will contain the new bookmark.  i.e. it can be from the origin.  This can be checked with `dsl_dataset_is_before()`, like we do for normal bookmark creation.  Or better yet, by using` dsl_bookmark_lookup()` which will call `dsl_dataset_is_before()` for us.

If this is not the case, the ioctl interface should be changed so that userland isn't required to say the exact same thing (the name of the filesystem) twice.  i.e. the value would omit the filesystem name.

problame replied:
> If this is not the case, the ioctl interface should be changed so that userland isn't required to say the exact same thing (the name of the filesystem) twice. i.e. the value would omit the filesystem name.

I think it is safer to leave as much string expansion logic as possible in userspace or a safe language like Lua, and have the C interface always expect full paths, albeit allowing the 'different source dataset' scenario you outlined above.

To be clear, given the plan to merge `dsl_bookmark_clone_*` and `dsl_bookmark_create_*` machinery, the nvlist that both zcp and ioctl will pass to `dsl_bookmark_create` will look like this (agreed?):

```
{
        // with source on origin
        "tank/fs1#bm1_clone" : "tank/fs1_origin#bm1",
        "tank/fs1#newbm_of_a" : "tank/fs1_origin@a",

        // with source on same dataset: still have full paths (expansion done by caller, see zfs_do_bookmark)
        "tank/fs1#b" : "tank/fs1@b",
        "tank/fs1#aboomkark_clone" : "tank/fs1#abookmark",
}
```

problame replied:
FYI: the proposal above is what's implemented in https://github.com/zfsonlinux/zfs/commit/fa99694ea55fb6a1477e2804a7e150bb31e3b84a see function `dsl_bookmark_create_nvl_validate`

problame commented:
Changes in the force-push:
- prettify commit history prior to addressing review round
- Merge `dsl_bookmark_clone` machinery into `dsl_bookmark_create`
- Implement [my](https://github.com/zfsonlinux/zfs/pull/9571#discussion_r365010763) proposal to always use full paths in the nvlists, both from ioctl and zcp
- Make `zfs bookmark ds#redactionbookmark ds#newbookmark` create a 'normal' bookmark that doesn't reference the redaction bookmark, put explicit notice in man page
  - @pcd1193182 is ok with that
  - if full copies of redaction bookmarks are desired at some point in the future, a CLI flag such as `zfs bookmark --redaction ds#redactionbookmark ds#newbookmark` could be used

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -954,7 +1155,9 @@ dsl_bookmark_destroy(nvlist_t *bmarks, nvlist_t *errors)
-	nvpair_t *pair = nvlist_next_nvpair(bmarks, NULL);
+	nvpair_t *pair;
+
+	pair = nvlist_next_nvpair(bmarks, NULL);
```
[nit] this seems like an unnecessary change.

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -432,12 +636,9 @@ dsl_bookmark_create_redacted_check(void *arg, dmu_tx_t *tx)
 
-	rv = dsl_dataset_hold(dp, dbcra->dbcra_snap,
-	    FTAG, &snapds);
 	if (rv == 0) {
```
no need for this check, `rv` is always zero at this point.

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -377,7 +500,78 @@ dsl_bookmark_create_sync_impl(const char *bookmark, const char *snapshot,
+	VERIFY0(dsl_bookmark_hold_ds(dp, new_name, &bmark_fs_new, FTAG,
+	    &new_shortname));
+	/* Bookmarks must be on the same dataset */
+	VERIFY3U(bmark_fs_source->ds_object, ==, bmark_fs_new->ds_object);
```
I think you should be able to create a bookmark from a bookmark in the origin, just like you can create a bookmark from a snapshot in the origin.

problame replied:
Yes, I already addressed this in the `*_check` but forgot to adjust the `*_sync`. Should I do the `dsl_dataset_is_before` check again in a VERIFY or just drop it?

problame replied:
... dropping it

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -175,25 +299,20 @@ dsl_bookmark_create_check(void *arg, dmu_tx_t *tx)
 	}
 
-	return (rv);
+	return (SET_ERROR(rv));
```
Normally SET_ERROR() appears where the error is first generated, where we see the EINVAL (or whatever) literal.  Is there a reason for doing it here?

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -116,57 +120,177 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
+		    pair2 != NULL; pair2 = nvlist_next_nvpair(bmarks, pair2)) {
+			if (strcmp(nvpair_name(pair), nvpair_name(pair2)) == 0)
+				return (-1);
+		}
```
Why is this needed?  We already do it in zfs_ioc_bookmark(), and the ZCP code only has one entry in the nvlist.

problame replied:
I think it makes sense to have _one_ function that validates the schema of the bookmark_create nvlist, and have that function be in the common code path for both ioctl and zcp (both ioctl and zcp call `dsl_bookmark_create_check`, which in turn calls this function).

I propose to remove the uniqueness check in the ioctl handler.
It shouldn't make a noticable performance difference for the zcp use case with a single-pair nvlist.

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -116,57 +120,177 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
+		return (error);
+
+	return (0);
+}
```
I think we could get rid of these 2 functions and put their code in-line.  They're only a couple lines of code.

problame commented:
- [ ] naming: do we want to name the thing bookmark "cloning", "copying", ...? (would require another build)
- [ ] test case for bookmark copying with source on different fs that is origin of newbookmark's fs
- [ ] test / review empty nvlist
- [ ] after the above: another review round

problame commented:
The `ubuntu 16.04 TEST` result link is a redirection loop.

The CentOS 7 failure:
```

Tests with results other than PASS that are unexpected:
    SKIP cli_root/zpool_create/zpool_create_008_pos (expected PASS)
```
doesn't really seem to be a failure (I did not mark it as SKIP anywhere)

The coverage test fails in ztest, but I cannot reproduce the crashes on my machine.

@ahrens  @pcd1193182 I think this PR is ripe for another round of code review. See the checkboxes in prior comments for what changed since the last one. Thanks!

behlendorf commented:
> The ubuntu 16.04 TEST result link is a redirection loop.

Sorry about that, it was a CI related infrastructure failure was has now been sorted out for the next run.

> The CentOS 7 failure:

Right, unrelated and recently fixed in master by e5030fbc2846c347f58e3a0bae436da84b98ef16.

> The coverage test fails in ztest, but I cannot reproduce the crashes on my machine.

The ztest failures look to be unrelated, but the reported [memory leak](http://build.zfsonlinux.org/builders/Ubuntu%2018.04%20x86_64%20Coverage%20%28TEST%29/builds/325/steps/shell_9/logs/kmemleak) should be looked in to.

problame commented:
@behlendorf found the memory leak, added a reproducer / unit test for it. Queued up as a separate commit https://github.com/zfsonlinux/zfs/pull/9571/commits/668ad1ed86fd570ac62206d89ebfbbdfb348ea8b , maybe consider cherry-picking it to 0.8.3 (cc @pcd1193182  )

problame commented:
Sorry, had to fix that typo...

@behlendorf found the memory leak, added a reproducer / unit test for it. Queued up as a separate commit b3c433ed80047be89a271d88f4313dd1d677535b , maybe consider cherry-picking it to 0.8.3 (cc @pcd1193182 )

problame commented:
- [x] rebased onto origin/master, should fix CentOS build failure

problame commented:
In order to avoid the need to unnecessarily squash this PR down to 5 commits just to make buildbot run tests on all commits, I opened #9867 
Reviewers of this PR can hence ignore the first three commits (they are small anyways).

behlendorf commented:
@problame it looks like several commits in this stack could each be moved to their own PRs.  This would help me get them merged more quickly.  Specifically, I'm thinking of:

f72eddc dsl_bookmark_create_check: fix NULL pointer deref if dbca_errors == NULL
8c6a30f zfs_namecheck: entity_namecheck: update doc comment to include space as allowed
c1eb601 scripts/zfs-test.sh: example for -t

Also #9867 has been merged.

problame commented:
> @problame it looks like several commits in this stack could each be moved to their own PRs. This would help me get them merged more quickly. Specifically, I'm thinking of:
> 
> [f72eddc](https://github.com/zfsonlinux/zfs/commit/f72eddc5839c1dc93bc728012f67c996448b083e) dsl_bookmark_create_check: fix NULL pointer deref if dbca_errors == NULL
> [8c6a30f](https://github.com/zfsonlinux/zfs/commit/8c6a30f233bc48b557e0d6aee5dfad14b7bfa88e) zfs_namecheck: entity_namecheck: update doc comment to include space as allowed
> [c1eb601](https://github.com/zfsonlinux/zfs/commit/c1eb60134c2cfce58f2fe91d6536754d787b5a93) scripts/zfs-test.sh: example for -t
> 
> Also #9867 has been merged.

Opened the respective PRs, one for each commit.
I have rebased this branch onto latest master and resolved conflicts with the zcp inherit support, but won't force-push until those one-commit PRs are merged and this branch is rebased again.

behlendorf commented:
@problame thanks for splitting out those small changes.  They've now been merged.

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -116,84 +120,206 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
+	switch (error) {
+	case ESRCH:
+		/* happy path: new bmark doesn't exist, proceed after switch */
+		break;
```
let's explicitly set `error = 0` here, so that we aren't relying on later code to reset it.

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -116,84 +120,206 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
+	/* error is retval of the following if-cascade */
+	if (strchr(source, '@') != NULL) {
+		dsl_dataset_t *source_snap_ds;
+		ASSERT3B(snapshot_namecheck(source, NULL, NULL), ==, 0);
```
snapshot_namecheck doesn't return a boolean, so I think this should be `ASSERT3S`.

ahrens commented on module/zfs/dsl_bookmark.c:
```diff
@@ -381,7 +511,76 @@ dsl_bookmark_create_sync_impl(const char *bookmark, const char *snapshot,
+	    &new_shortname));
+
+	/*
+	 * create a copy of the source bookmark by coping most of its members
```
typo: `coping` -> `copying`

ahrens commented on module/zfs/dsl_bookmark.c:
typo: Bookmoark

problame commented on module/zfs/dsl_bookmark.c:
```diff
@@ -116,84 +120,206 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
+ * - each pair passes dsl_bookmark_create_nvl_validate_pair
+ * - all newbookmarks are in the same pool
+ * - all newbookmarks have unique names
+ * - FIXME: need to check that newbookmark is never used as source in any other?
```
I guess

```
fs#a -> fs@a
fs#b -> fs#a
```
would work because we run the list top-to-bottom.

```
fs#b -> fs#a
fs#a -> fs@a
```
would fail for the first pair with an error indicating `fs#a` does not exist, and have success with the second pair.

Not sure if we want to commit to the implementation detail that the list is executed top-to-bottom, and not in parallel.

What do you think?

ahrens replied:
I agree it would be nice to disallow this.  We could check that all the sources exist, before creating any of them.

problame replied:
`dsl_bookmark_create_nvl_validate`'s job is schema validation.

`dsl_bookmark_create_check` and `dsl_bookmark_create_redacted_check` already check existence for all sources.

I'll update the documentation of `dsl_bookmark_create_nvl_validate` and add the check to prevent the behavior described above.

ahrens replied:
I think it would be fine for "schema validation" to not check that the sources exist (and not care about if they are trying to create the bookmark and use it in the same operation), just that their names are allowed strings, and they are all in the same pool.

problame replied:
OK, so if I understand correctly, the current behavior is fine, i.e. it is ok that 
```
fs#a -> fs@a
fs#b -> fs#a
```
will create both bookmarks?

problame replied:
- [x] extend pyzfs unittests for bookmark creation with missing sources
  to explicitly check that no bookmarks were created if a source is missing

problame commented on module/zfs/dsl_bookmark.c:
```diff
@@ -116,84 +120,206 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
+			return (-1);
+
+		if (*(bmark + first_len) != '/')
+			return (-1);
```
will add a test case for that

problame commented:
- [x] fix bookmark creation of snapshots of the pool dataset
    - [x] add corresponding regression test
- [x] extend pyzfs unittests for bookmark creation with missing sources
  to explicitly check that no bookmarks were created if a source is missing
- [x] more concise doc comments
- [x] use correct assert macro
- [x] address spacecheck todos in zcp code
- [x] fix typos

problame commented:
Something about the codecov upload / processing seems to be broken again (cc @behlendorf )

Apart from that, all pending review remarks have been addressed.

pcd1193182 commented on module/zcommon/zfs_namecheck.c:
```diff
@@ -183,6 +183,8 @@ entity_namecheck(const char *path, namecheck_err_t *why, char *what)
 {
 	const char *end;
 
+	ASSERT((why == NULL && what == NULL) || (why != NULL && what != NULL));
```
You can use `EQUIV(why == NULL, what == NULL)` for this, and it'll give a better error message

pcd1193182 commented on module/zfs/dsl_bookmark.c:
```diff
@@ -116,84 +120,209 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
+		error = SET_ERROR(EEXIST);
+		goto eholdnewbmds;
+	default:
+		error = SET_ERROR(error); /* dsl_bookmark_lookup_impl doesn't */
```
Isn't that because the zap_lookup code does?

problame replied:
apparently, I didn't look that deep into the call graph when I first learned about `SET_ERROR`

pcd1193182 commented on tests/zfs-tests/tests/functional/cli_root/zfs_bookmark/zfs_bookmark_cliargs.ksh:
```diff
@@ -97,4 +119,129 @@ log_mustnot zfs bookmark "$TESTSNAP" "$DATASET#"
+log_must zfs bookmark "$DATASET#$TESTBM" "#$TESTBMCOPY"
+## if FUTURE_WORK; then # try to use the copy even after the source bookmark is destroyed
+# log_must zfs destroy "$DATASET#$TESTBM"
+# log_must eval "zfs send --redact "$TESTBMCOPY" $DATASET@$TESTSNAP > $TEST_BASE_DIR/zfstest_datastream.$$"
```
Do we want to leave these commented out lines in here? We can always readd them later when the future work is done, since they're in the commit history on github.

problame replied:
If we remove them in this PR, we squash them away, so they won't be available in the future.

problame replied:
I'll delete them, the big comment in `dsl_bookmark.c` that explains why we don't support deep-copying of redaction bookmarks should be sufficient for anyone who starts working on this in the future.

problame commented:
Addressed @pcd1193182 's remarks:
- [x] `EQUIV`
- [x] remove redundant `SET_ERROR(error) /* dsl_bookmark_lookup_impl doesn't */`
- [x] drop "future work" comments in test suite

problame commented:
The CI failure seems like a flaky test.

ahrens commented on contrib/pyzfs/libzfs_core/_constants.py:
```diff
@@ -22,11 +22,15 @@
 
 
 # https://stackoverflow.com/a/1695250
```
I realize that this is "just" `contrib` code, but it would be preferable for the comment to explain the code.  We could consider also linking to an external expanded explanation if needed, while weighing the risk of this link becoming stale.  I also realize you aren't adding this link, but if you have time, fixing this would be appreciated.

We could explain why we can't use the built-in python enums (presumably because we want to support pre-2.4 python installs); explain approximately what this code does (creates a new type?); and that this is the canonical way of doing so on old python.  It may then not be necessary to include the external link.

problame replied:
I just ~~cargo-culted~~ generalized the existing enum type constructor/generator to do what I needed it to do.
I'd guess the intention behind the link was to point a future developer to practical examples for enum type construction.
FWIW: The stackoverflow page is archived on archive.org ¯\_(ツ)_/¯

I figure future work should replace `_constants.py` with code-generation anyways, so I guess this is not a blocker for merging this feature.

behlendorf replied:
This potentially looks like a good candidate to cleanup in another PR if someone is interested.  Python 2 officially [went EOL in January 2020](https://www.python.org/doc/sunset-python-2/) so bumping the minimum supported Python version to 3.4 makes sense.  I tend to agree this doesn't need to block this feature.

behlendorf commented:
Merged as:

948f0c4 zcp: add zfs.sync.bookmark
a73f361 Implement bookmark copying
